runtime/v2/shim/util_windows.go

```diff
@@ -57,35 +57,31 @@ func SocketAddress(ctx context.Context, id string) (string, error) {
 
 // AnonDialer returns a dialer for a npipe
 func AnonDialer(address string, timeout time.Duration) (net.Conn, error) {
-	var c net.Conn
-	var lastError error
-	timedOutError := errors.Errorf("timed out waiting for npipe %s", address)
 	start := time.Now()
 	for {
 		remaining := timeout - time.Since(start)
 		if remaining <= 0 {
-			lastError = timedOutError
-			break
+			return nil, errors.Errorf("timed out waiting for npipe %s", address)
 		}
-		c, lastError = winio.DialPipe(address, &remaining)
-		if lastError == nil {
-			break
+		c, err := winio.DialPipe(address, &remaining)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return nil, err
+			}
+			// There is nobody serving the pipe. We limit the timeout for this case
+			// to 5 seconds because any shim that would serve this endpoint should
+			// serve it within 5 seconds. We use the passed in timeout for the
+			// `DialPipe` timeout if the pipe exists however to give the pipe time
+			// to `Accept` the connection.
+			// Return `err` here as it will be ERROR_FILE_NOT_FOUND.
+			if time.Since(start) >= 5*time.Second {
+				return nil, err
```

Comment:
Could be useful to add an `errors.Wrap` here that a timeout occurred. This would still give access to the underlying error for the caller to check. Generally we discourage direct checking as `if err == ERR_SOMETHING` , in favor of `if IsSomething(err)` which does the `errors.Cause` unwrapping anyway.

Reply:
Added wrapping.